proposals/0000-extending-capi-calling-convention.rst

```diff
@@ -0,0 +1,96 @@
+
+For category 2, the parameter is passed by the value of the encase primitive like category 1. This category corrisponds with C structs that possess only one member of a primitive type (example: 
+``typedef struct { int value } unique_id;``). In the C wrapper function, the corrisponding parameter has the member type and is encapsulated in a struct before being passed to the wrapped function.
+For return value in this category, the value is returned in the primitive type of the singleton member. 
+
+Parameters of category 3 are passed by pointers. The wrapper function shalls dereference these pointers before passing them to the wrapped function.  
+If the return value is in this category, the wrapper function shalls have one extra rightmost pointer parameter where the result of the wrapped function will be written to. 
+The types of category 3 must implement a Storable instance in order for Haskell to marshal them to and from pointers. 
+
+Examples
+--------
+Suppose the the C function ``aggregate wrapped(int, aggregate)`` is needed to be called from Haskell. This would be how it is done. 
+
+    data {-# CTYPE "aggregate" #-} Aggregate = ...
+    instance Storable Aggregate where ...
+ 
+    foreign unsafe capi "header.h wrapped" ffunc :: CInt -> Aggregate -> Aggregate
+
+Generated C wrapper function
+
+    void wrapper (int _0, aggregate \*_1, aggregate \*_result) { \*_result = wrapped(_0, \*_1); }
+
+Equivalent in ccall convention
+
+    foreign unsafe ccall "wrapper" ffunc_wrapper :: CInt -> Ptr Aggregate -> Ptr Aggregate -> IO ()
+    ffunc :: CInt -> Aggregate -> Aggregate
+    ffunc _0 _1 = unsafePerformIO (with _1 (\ _1' -> alloca (\ _result -> ffunc_wrapper _0 _1 _result >> peek _result)))
+
+
+For cases of category 2
+
+    newtype {-# CTYPE "unique_id" #-} UniqueId = UniqueId CInt
+    
+    foreign unsafe capi "header.h wrapped" ffunc :: UniqueId -> Ptr () -> IO UniqueId 
+
+C wrapper function
+
+    int wrapper (int _0, void \*_1) { unique_id result = wrapped((unique_id){_0}, _1); return *(int*)&result; }
+
+ccall equivalent
+
+    foreign unsafe ccall "wrapper" ffunc_wrapper :: UniqueId -> Ptr () -> IO UniqueId
+    ffunc :: UniqueId -> Ptr () -> IO UniqueId
+    ffunc _0 _1 = ffunc_wrapper _0 _1
+
+Effect and Interactions
+-----------------------
+The modification shoud make interfacing with C code easier, especially for external C libraries.
+
+Costs and Drawbacks
+-------------------
+The cost is limited to the initial implementation. Little maintenance should be required due to the simplicity of the changes.
+
+Backward Compatibility
+----------------------
+Since this proposal only affect capi call that previously would have been illegal, it is fully backward compatible.
+
+Alternatives
+------------
+Another approach is to flatten structs into their primitive members and reconstruct the structs inside the wrapper function. 
+This grant minor performance benefit but is more difficult to implement and does not for unions.
```

Comment:
```suggestion
Another approach is to flatten structs into their primitive members and reconstruct the structs inside the wrapper function. 
This would grant a minor performance benefit but is more difficult to implement and does not work for unions.
```